src/lib.rs

```diff
@@ -56,6 +70,85 @@ fn _byte_pair_merge(
     parts
 }
 
+fn _byte_pair_merge_large(
+    ranks: &HashMap<Vec<u8>, Rank>,
+    piece: &[u8],
+) -> Vec<(usize, Rank)> {
+    let get_rank = |start_idx: usize, end_idx: usize| {
+        *piece.get(start_idx..end_idx)
+            .filter(|p| p.len() < piece.len())
+            .and_then(|p| ranks.get(p))
+            .unwrap_or(&Rank::MAX)
+    };
+
+    let mut rank_indexes = BTreeMap::<Rank, BTreeSet<usize>>::new();
+    let mut index_rank = vec![Rank::MAX; piece.len() + 1];
```

Comment:
mutations seemed easier this way, compared to creating a struct with index/rank/prev/next